### PR TITLE
fix #231

### LINKED ISSUE #231: sdl-triangle example uses 100% CPU in some window managers
Nvidia 1060, fvwm2

CC @int-e

### PR DESCRIPTION
**Pull request by int-e**
Before this patch, `sdl-triangle`'s main loop would redraw continuously
in a tight loop, consuming 100% CPU on my system and rendering X11
unresponsive. The solution is to wait for an event instead.

### CODE REVIEW

**Comment by dpwiz** [comment]
There should be a proper frame-limiting gadget. Having this in examples introduces a gotcha when you move from trivial example to a slightly more complicated non-interactive animation.

**Comment by expipiplus1** [comment]
Thanks!

I don't see how it was redrawing so fast though in FIFO mode.

> There should be a proper frame-limiting gadget. Having this in examples introduces a gotcha when you move from trivial example to a slightly more complicated non-interactive animation.

Yeah, for sure.

**Comment by expipiplus1** [comment]
hmm, resizes aren't events, so this causes too-small rendering when the window is enlarged. I've fixed in another branch though by adding a timeout,

**Comment by expipiplus1** [comment]
Still some artifacts though. but at least on my machine moving the window and resizing it are much smoother.

**Comment by int-e** [comment]
Works for me? Also, there *are* resize events:

  https://hackage.haskell.org/package/sdl2-2.5.2.0/docs/SDL-Event.html#t:WindowResizedEventData

Could there be an effect of double buffering (which I didn't properly think about), causing an old frame to be shown?

There's also timers, of course... though they're based on callbacks so state would have to be managed in MVars (or IORefs)

  https://hackage.haskell.org/package/sdl2-2.5.2.0/docs/SDL-Time.html#v:addTimer

**Comment by int-e** [comment]
So I've tried dumping events (there are resize events for me). But there's an interesting validation warning in this case, which seems in line with attempting to work with outdated frames:
```
Event {eventTimestamp = 14175, eventPayload = WindowSizeChangedEvent (WindowSizeChangedEventData {windowSizeChangedEventWindow = Window 0x0000000004b44bd0, windowSizeChangedEventSize = V2 1602 451})}
Event {eventTimestamp = 14175, eventPayload = WindowResizedEvent (WindowResizedEventData {windowResizedEventWindow = Window 0x0000000004b44bd0, windowResizedEventSize = V2 1602 451})}
Event {eventTimestamp = 14175, eventPayload = WindowExposedEvent (WindowExposedEventData {windowExposedEventWindow = Window 0x0000000004b44bd0})}
Validation: Validation Warning: [ UNASSIGNED-BestPractices-Error-Result ] Object 0: handle = 0x4b30b00, type = VK_OBJECT_TYPE_INSTANCE; | MessageID = 0x5f379b89 | vkQueuePresentKHR(): Returned error VK_ERROR_OUT_OF_DATE_KHR.
Using present mode PRESENT_MODE_FIFO_KHR
```
I don't understand the program logic well enough (yet?) to try to fix this.

**Comment by expipiplus1** [comment]
OK, I think I see what's going on here. The resize event does come in and we start to render a frame (with the old size since we don't know better. 

Since we have the wrong size, when we try to acquire the image we can't and the program aborts that render and recreates the swapchain. No more events come in  (until we move the mouse or something) and the new swapchain isn't used.

That validation warnings is bogus btw: https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/2335

**Comment by expipiplus1** [comment]
This also revealed a bug in resize:

- Start rendering a frame, some command buffer using a descriptor set is in flight now
- Resize the window and invalidate the swapchain
- `acquireNextImageKHR` succeeds without waiting, and we get an an index to a new image.
- `updateDescriptorSets` tries to update the descriptor set at that index, but that set is still in use by the in flight command buffer.

Basically access to the descriptor sets is valid with the assumption that there can't be more than one frame in flight with the same index. I think that this change just exposed this race, not caused it.
